fennel/testing/execute_aggregation.py

```diff
@@ -326,47 +329,240 @@ def get_val(self) -> List:
+                ].min(axis=1)
+            input_df = pd.concat([input_df, del_df], ignore_index=True)
+        else:
+            # forever hopping aggregates on keyed datasets
+            if FENNEL_DELETE_TIMESTAMP in input_df.columns:
+                del_df = input_df.copy()
+                del_df = del_df[del_df[FENNEL_DELETE_TIMESTAMP].notna()]
+                del_df[FENNEL_ROW_TYPE] = -1
+                del_df[ts_field] = del_df[FENNEL_DELETE_TIMESTAMP]
+                input_df = pd.concat([input_df, del_df], ignore_index=True)
+
+        input_df = input_df.sort_values(
+            [ts_field, FENNEL_ROW_TYPE], ascending=[True, False]
+        )
+        return input_df
+    elif isinstance(aggregate.window, Session):
+        gap = aggregate.window.gap_total_seconds()
+        cols = input_df.columns
+        input_df = (
+            (
+                input_df.groupby(key_fields)
+                .apply(
+                    get_timestamps_for_session_window,
+                    key_fields=key_fields,
+                    ts_field=ts_field,
+                    gap=gap,
+                )
+                .reset_index()
+            )
+            .loc[:, cols]
+            .assign(
+                **{
+                    ts_field: lambda x: x[ts_field].astype(
+                        pd.ArrowDtype(pa.timestamp("ns", "UTC"))
+                    )
+                }
+            )
         )
-        # For every row find the time it will expire and add it to the dataframe
-        # as a delete row
-        del_df = df.copy()
-        del_df[ts_field] = del_df[ts_field] + expire_secs
+
+        # Marking the delete timestamp as end timestamp of next session window.
+        fennel_internal_field = "__fennel_internal_field__"
+        internal_df = input_df.copy()[key_fields + [ts_field]].drop_duplicates()
+        internal_df[fennel_internal_field] = internal_df.groupby(key_fields)[
+            ts_field
+        ].shift(-1)
+        internal_df = internal_df[internal_df[fennel_internal_field].notna()]
+
+        del_df = input_df.copy().merge(
+            internal_df, how="inner", on=key_fields + [ts_field]
+        )
+        del_df[ts_field] = del_df[fennel_internal_field]
+        del_df.drop(columns=[fennel_internal_field], inplace=True)
         del_df[FENNEL_ROW_TYPE] = -1
         # If the input dataframe has a delete timestamp field, pick the minimum
         # of the two timestamps
-        if FENNEL_DELETE_TIMESTAMP in df.columns:
+        if FENNEL_DELETE_TIMESTAMP in input_df.columns:
             del_df[ts_field] = del_df[[ts_field, FENNEL_DELETE_TIMESTAMP]].min(
                 axis=1
             )
-        df = pd.concat([df, del_df], ignore_index=True)
-        df = df.sort_values(
+        input_df = pd.concat([input_df, del_df], ignore_index=True)
+        input_df = input_df.sort_values(
             [ts_field, FENNEL_ROW_TYPE], ascending=[True, False]
         )
```

Comment:
Lets move the common part of sorting and combining deletes out

Reply:
moved